---
title: "R Notebook"
output:
  html_document: default
  html_notebook: default
---

## Introduction

I've created 1,000 simulations of populations undergoing different levels of 
sexual reproduction and analyzed the index of association over 10,000 loci. I 
have analyzed the index of association for 10, 25, 50, and 100 samples from 
each population and analyzed $\bar{r}_d$ for the data set with different levels
of missing data: 1%, 5%, and 10%

## Setup

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width = 10, fig.height = 10)
```



```{r, required_packages}
library('zksimanalysis')
```

## Data retrieval

The data were downloaded from the CGRB server using rsync with the following
command.

> Note: replace XXXXX with the port number. I am not writing it here for
security reasons.

```
rsync --update -avz -e "ssh -p XXXXX" --exclude 'twenty_loci_results' \ 
kamvarz@files.cgrb.oregonstate.edu:/nfs1/BPP/Grunwald_Lab/home/kamvarz/simulation_analysis/results \ 
../simulation_results
```


```{r, load_data, results = "hide"}
# Here we have to filter out the stored genetic data
genomic_data <- load("../../simulation_results/genomic_data.rda")
genomic_data
```

## Visualizations

```{r, themes}
old_theme <- theme_set(theme_bw())
old_theme <- theme_update(axis.text.x = element_text(angle = 90, vjust = 0.5))
old_theme <- theme_update(text = element_text(size = 14))
asp       <- function(r) theme(aspect.ratio = r)
# The p-value for rd is best visualized on a log scale. This provides that
# the color scaling for that transformation, emphasizing smaller p-values over
# larger ones.
log_rd_color <- function(breaks = c(0.005, 0.01, 0.025, 0.05, 0.1)){
  scale_color_viridis(option = "viridis", 
                      breaks = log(breaks), 
                      labels = breaks)
}
  
# Various labels with mathematical expressions are used for the plots. 
# Pre-defining them here allows us to use eval(exp) instead of re-typing the
# entire expression every time.
rd    <- quote(expression(bar(r)[d]))
rdcc  <- quote(expression(paste(bar(r)[d], " (clone-corrected)")))
prd   <- quote(expression(paste("p-value (", bar(r)[d], ")")))
prdcc <- quote(expression(paste("p-value (", bar(r)[d], " (clone-corrected))")))
E5g   <- quote(expression(paste("Genotypic evenness (", E[5][G], ")")))
E5a   <- quote(expression(paste("Mean allelic evenness (", E[5][A], ")")))
Hexp  <- quote(expression(paste("Nei's gene diversity (", H[exp], ")")))


fancy_clone_correct <- . %>% 
  mutate(clone_correct = factor(clone_correct, labels = c("uncorrected", "clone-corrected")))
fancy_sample <- . %>%  
  mutate(sample = forcats::lvls_revalue(sample, paste("n =", sort(unique(sample)))))
fancy_mutation_rate <- . %>% 
  mutate(mutation_rate = factor(mutation_rate, labels = c("Even Mutation Rate", "Uneven Mutation Rate")))
  
```


Visualizations are one of the most important steps in an analysis. Here's why.
Below are the results when the mutation rate is 1e-5

```{r, fig.width = 6, fig.height = 5}
missing_plot <- missing_stats %>% 
  select(sexrate, rd, sample, matching, missing) %>%
  fancy_sample %>%
  mutate(missing = paste0(sprintf("%d", missing*100), "%")) %>%
  mutate(missing = factor(missing, levels = unique(missing))) %>%
  mutate(matching = ifelse(matching == "match", "ignored", "new allele")) %>%
  ggplot(aes(x = sexrate, y = rd, fill = matching)) +
  geom_violin(scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) +
  scale_fill_grey(start = 1, end = 0.5) +
  facet_grid(sample ~ missing, switch = "y") +
  scale_y_continuous(position = "right") +
  theme(text = element_text(size = 14)) +
  theme(strip.text = element_text(size = 14, face = "bold")) +
  theme(strip.background = element_blank()) +
  theme(aspect.ratio = 1/2) +
  theme(legend.position = "bottom") +
  labs(list(
    # title = "index of association across missing data",
    x = "Rate of Sexual Reproduction",
    y = eval(rd),
    fill = "Missing Data"
    ))

ggsave(filename = "~/Documents/Grunwald/dissertation/figure/simulations/genomic_missing.pdf",
       plot = missing_plot, dpi = 600, height = 5, width = 6)

main_rd_plot <- ggplot(original %>% select(sexrate, rd, sample), 
       aes(x = sexrate, y = rd, fill = sample)) +
  geom_boxplot() +
    labs(list(
    title = "index of association by sex rate",
    x = "Rate of Sexual Reproduction",
    y = eval(rd),
    fill = "Sample Size"
    ))

pval_rd_plot <- ggplot(original %>% select(sexrate, p.rD, sample), 
       aes(x = sexrate, y = p.rD, fill = sample)) +
  geom_boxplot() + 
    labs(list(
    title = "p-value for the index of association by sex rate",
    subtitle = "out of 999 permutations",
    x = "Rate of Sexual Reproduction",
    y = eval(prd),
    fill = "Sample Size"
    ))

main_rd_plot
pval_rd_plot
missing_plot

```



This is quite weird. If you look at the value for the index of association, we
can see that the minimum value is about 0.02. To get a better sense of what's
going on, we need to take a look at the value of the null distributions


```{r, distplot, fig.width = 10, fig.height = 10}
null_dist <- ggplot(original %>% unnest(samples) %>% select(samples, sample, sexrate), 
       aes(x = samples)) +
  geom_density(aes(fill = sexrate), alpha = 0.5) +
  scale_fill_viridis(discrete = TRUE) +
  facet_wrap(~sample, ncol = 1, scale = "free_y") +
  scale_x_continuous(limits = c(-5e-4, 1e-3)) +
  ggtitle("Null distributions for the index of assocation by sample size and sex rate")
```

An interesting thing to note is that it seems as if, with greater linkage, the
null distributions are more dependent on sample size.

## Explanation of weird phenomena

These are narrow, but it's not like they don't have variation, so there must be
something else going on. To illustrate, I'll load a clonal data set and a sexual
data set. 

```{r, load_examples}
load("../../simulation_results/genomic_rda_files/X.data.3639601.100.bpp.dir_genomic10_gen_10seed_0_sex_1.0000_gen_10000_rep_09.pop.feather.DATA.rda") %>% 
  get() %>% 
  tidyr::gather(key = percent_missing, value = dataset, starts_with("dataset")) -> x

xdat <- x$dataset[[2]]
xdat
load("../../simulation_results/genomic_rda_files/X.data.3639601.1.grungall1.dir_genomic10_gen_10seed_0_sex_0.0000_gen_10000_rep_01.pop.feather.DATA.rda") %>%
  get() %>%
  tidyr::gather(key = percent_missing, value = dataset, starts_with("dataset")) -> y

ydat <- y$dataset[[2]]
ydat
```

Now, I'll simulate a sexual population using the `glSim()` functin from *adegenet*:

```{r, glsim}
set.seed(20161018)
xsim <- glSim(100, 1e4, 0, ploidy = 2) %>% as.snpclone()
xsim
```

Now, let's look at the values of $\bar{r}_d$:

```{r, datrd}
barplot(c(sexual = bitwise.ia(xdat), 
          clonal = bitwise.ia(ydat),
          simulated = bitwise.ia(xsim)), log = "y", ylab = "log index of association")
```

We can clearly see that the value of $\bar{r}_d$ is different between the sexual
and simulated population. What's causing this difference? A quick look at the
dendrograms produced will give us a good idea:

```{r, dendrogrammin}
library('ape')
xdat %>% bitwise.dist() %>% nj() -> sexdist
ydat %>% bitwise.dist() %>% nj() -> clonedist
xsim %>% bitwise.dist() %>% nj() -> simdist
par(mfrow = c(1, 3))
plot(sexdist, main = "Sexual")
add.scale.bar(x = 0.1, y = -0.1, lcol = "red", length = 0.05)
plot(clonedist, main = "Clonal")
add.scale.bar(x = 0.1, y = -0.1, lcol = "red", length = 0.05)
plot(simdist,   main = "Random (adegenet::glSim)")
add.scale.bar(x = 0.05, y = -0.1, lcol = "red", length = 0.05)
par(mfrow = c(1, 1))
```


Okay, that clears things up a little bit. The simulated population has a LOT of
variance. When happens to our populations if we shuffle them?

```{r, dendrogrammin_2_electric_boogaloo}
xdat %>% shuffle_genlight(mat = as.matrix(.), x = ., blocksize = 1) %>% bitwise.dist() %>% nj() -> sexdist
ydat %>% shuffle_genlight(mat = as.matrix(.), x = ., blocksize = 1) %>% bitwise.dist() %>% nj() -> clonedist
xsim %>% shuffle_genlight(mat = as.matrix(.), x = ., blocksize = 1) %>% bitwise.dist() %>% nj() -> simdist
par(mfrow = c(1, 3))
plot(sexdist, main = "Sexual")
add.scale.bar(x = 0.05, y = -0.1, lcol = "red", length = 0.05)
plot(clonedist, main = "Clonal")
add.scale.bar(x = 0.05, y = -0.1, lcol = "red", length = 0.05)
plot(simdist,   main = "Random (adegenet::glSim)")
add.scale.bar(x = 0.05, y = -0.1, lcol = "red", length = 0.05)
par(mfrow = c(1, 1))
```

One question I still have is: what does it look like with under prinicpal
component analysis. The goal of this is to find the variables that contribute to
the separation of the data.

```{r, prcomponents, cache = TRUE}
xdat %>% glPca(nf = 100) -> xdat.pc
ydat %>% glPca(nf = 100) -> ydat.pc
xsim %>% glPca(nf = 100) -> xsim.pc

xdat.s <- xdat.pc$scores[, 1:2]
ydat.s <- ydat.pc$scores[, 1:2]
xsim.s <- xsim.pc$scores[, 1:2]
lims <- range(c(xdat.s, ydat.s, xsim.s))
plot(xdat.s, xlim = lims, ylim = lims, col = "red")
points(ydat.s, xlim = lims, ylim = lims, col = "blue")
points(xsim.s, xlim = lims, ylim = lims)
legend("topright", legend = c("Sexual", "Clonal", "Random (adegenet::glSim)"), 
       col = c("red", "blue", "black"), pch = c(1, 1, 1))
```

### Re-analyze index of association


Since we saw all significant values for $\bar{r}_d$ based on randomly permuting
linked markers, we'll test here what happens when we randomly permute the
chromosomes. This was done on the OSU CGRB cluster and downloaded here.

Since not all of the replicates from the simulations finished correctly, we are
sampling the minimum number of working simulations from the data per seed, which
is 7. The dataset is called `original_redux`

```{r, index_table, resuts = "asis"}
original_redux %>% 
  group_by(sexrate) %>%
  summarize(n = n()/4)
original_redux %>% 
  group_by(sexrate, run) %>% 
  summarize(n = n()/4) %>% 
  spread(key = run, value = n)
```


```{r, fig.height=4, fig.width=6}
corrected_rd <- original_redux %>% 
  select(sexrate, rd, sample) %>%
  fancy_sample %>%
  ggplot(aes(x = sexrate, y = rd)) +
  geom_violin(scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) +
  geom_point(alpha = 0.25) +
  # geom_boxplot(outlier.color = "grey50", outlier.size = 0.5) + 
  # scale_fill_grey(start = 0.5, end = 1) + 
    labs(list(
    # title = "index of association by sex rate",
    x = "Rate of Sexual Reproduction",
    y = eval(rd)
    )) +
  facet_wrap(~sample) +
  theme(aspect.ratio = 1/2) +
  theme(strip.background = element_blank()) +
  theme(strip.text = element_text(size = 14, face = "bold")) +
  theme(text = element_text(size = 14)) 
corrected_rd

ggsave(filename = "~/Documents/Grunwald/dissertation/figure/simulations/genomic_rd.pdf",
       plot = corrected_rd, dpi = 600, height = 4, width = 6)


correcte_rd_p <- original_redux %>% 
  select(sexrate, p.rD, sample) %>%
  fancy_sample %>%
  ggplot(aes(x = sexrate, y = p.rD)) +
  geom_boxplot(outlier.color = "grey50", outlier.size = 0.5) +
  # geom_violin(scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) +
  scale_y_log10(breaks = c(0.01, 0.1, 1),
                minor_breaks = c((1:9)*0.001, (2:9)*0.01, (2:9)*0.1)) +
  # scale_fill_grey(start = 0.5, end = 1) + 
    labs(list(
    # title = "p-value for the index of association by sex rate",
    # subtitle = "out of 999 permutations",
    x = "Rate of Sexual Reproduction",
    y = eval(prd)
    # fill = "Sample Size"
    )) +
  facet_wrap(~sample) +
  theme(aspect.ratio = 1.618^-1) +
  theme(strip.background = element_blank()) +
  theme(strip.text = element_text(size = 14, face = "bold")) +
  theme(text = element_text(size = 14))
correcte_rd_p

```


Now we can take a look at the null distributions by sex rate and sample size

```{r, fig.height=7, fig.width=10}
no_y <- theme(axis.title.y = element_blank()) +
  theme(axis.text.y = element_blank()) +
  theme(axis.ticks.y = element_blank()) +
  theme(panel.grid.major.y = element_blank()) +
  theme(panel.grid.minor.y = element_blank()) 
null_dist <- original_redux %>% 
  unnest(samples) %>% 
  select(samples, sexrate, sample) %>%
  ggplot(aes(x = samples)) +
  geom_density(aes(fill = sexrate), alpha = 0.5) +
  scale_fill_viridis(discrete = TRUE, option = "plasma") +
  facet_wrap(~sample, ncol = 1, scale = "free_y") +
  labs(list(
    title = expression(paste("Null distributions of ", bar(r)[d])),
    subtitle = "For all sex rates",
    x = eval(rd)
  )) +
  no_y
null_dist

# Only the most clonal of the non-random mating
clone_dist <- original_redux %>% 
  unnest(samples) %>% 
  select(samples, sexrate, sample) %>%
  filter(sexrate %in% c("0.0000", "0.0001", "0.0005")) %>%
  ggplot(aes(x = samples)) +
  geom_density(aes(fill = sexrate), alpha = 0.5) +
  scale_fill_viridis(discrete = TRUE, option = "plasma") +
  facet_wrap(~sample, ncol = 1, scale = "free_y") +
  labs(list(
    title = expression(paste("Null distributions of ", bar(r)[d])),
    subtitle = "For sex rates < 0.1%",
    x = eval(rd)
  )) +
  no_y
clone_dist
```


## Session Information

```{r session_info}
options(width = 100)
devtools::session_info()
```

