---
title: "Post analysis of SSR simulation results"
output: html_document
---


## Introduction

I've created 10,000 simulations of populations undergoing different levels of
sexual reproduction and analyzed the index of association along with genotypic
and allelic diversity metrics, including an analysis of the contribution of each
locus to the genotypic diversity. This will serve as a report for the 
simulations.

## Setup

```{r, required_packages, message = FALSE}
library('zksimanalysis') # Main package for analysis
library('magrittr')      # For the %T>%
library('viridis')       # Excellent color scaling
library('dplyr')         # Magic data manipulation
library('tidyr')         # more magic
library('purrr')         # magic dealing with lists
library('ggplot2')       # magic plotting action!
```

This here will load in all of the results of the analyses. These were previously
processed in the file `ssr_data_cleaning.Rmd`. The name of the data was called
"vals".

```{r, load_data, cache = TRUE, results = "hide", cache.lazy = FALSE}
system.time({
res           <- load("../../simulation_results/processed_results.rda")
even_mutation <- vals %>% filter(mutation_rate == "even")
odd_mutation  <- vals %>% filter(mutation_rate == "uneven")
})
res
even_mutation
odd_mutation
```


All data coming in must be verified that we indeed have all data. Since we had a
factorial design, we can create what we expect the populations (records of the
simulations) to look like.

```{r, quality_control}
nrow(even_mutation) == 40000
nrow(odd_mutation) == 40000
```


## Visualizations

### Setup

The visualizations are summaries of the data and can help to understand the how
the data is distributed to see any patterns.

This part is setting up some helpful variables and names that we will use over
and over again.

```{r, plotting_params}
old_theme <- theme_set(theme_bw())
old_theme <- theme_update(axis.text.x = element_text(angle = 90, vjust = 0.5))
asp       <- function(r) theme(aspect.ratio = r)
# The p-value for rd is best visualized on a log scale. This provides that
# the color scaling for that transformation, emphasizing smaller p-values over
# larger ones.
log_rd_color <- scale_color_viridis(option = "viridis", 
                                    breaks = log(c(0.005, 0.01, 0.025, 0.05, 0.1)), 
                                    labels = c(0.005, 0.01, 0.025, 0.05, 0.1))
# Various labels with mathematical expressions are used for the plots. 
# Pre-defining them here allows us to use eval(exp) instead of re-typing the
# entire expression every time.
rd    <- quote(expression(bar(r)[d]))
rdcc  <- quote(expression(paste(bar(r)[d], " (clone-corrected)")))
prd   <- quote(expression(paste("p-value (", bar(r)[d], ")")))
prdcc <- quote(expression(paste("p-value (", bar(r)[d], " (clone-corrected))")))
E5g   <- quote(expression(paste("Genotypic evenness (", E[5][G], ")")))
E5a   <- quote(expression(paste("Mean allelic evenness (", E[5][A], ")")))
Hexp  <- quote(expression(paste("Nei's gene diversity (", H[exp], ")")))
```



```{r, rd_sexrate, fig.width = 10}
ggplot(vals, aes(x = sexrate, y = rbarD, fill = sample)) +
  geom_boxplot(outlier.size = 0.25) +
  facet_wrap(~mutation_rate) +
  labs(list(
    x = "Rate of Sexual Reproduction",
    y = eval(rd),
    title = "Standardized Index of Association vs Sex",
    subtitle = "for even and uneven mutation rates")
    ) 
```


```{r, rd_pval, fig.width = 10}
ggplot(vals, aes(x = sexrate, y = p.rD, fill = sample)) +
  geom_boxplot(outlier.size = 0.25) +
  scale_y_log10(breaks = c(0.01, 0.1, 1),
                minor_breaks = c((1:9)*0.001, (2:9)*0.01, (2:9)*0.1)) +
  facet_wrap(~mutation_rate) +
  labs(list(
    x = "Rate of Sexual Reproduction",
    y = eval(prd),
    title = "Detection of clonal reproduction",
    subtitle = "for even and uneven mutation rates")
    )
```

```{r, rd_pval_cc, fig.width = 10}
ggplot(vals, aes(x = p.rDcc, y = p.rD, pch = mutation_rate)) +
  geom_point(alpha = 0.25) +
  geom_abline(slope = 1) +
  scale_x_log10() +
  scale_y_log10() +
  facet_grid(sample~sexrate) +
  ylab(eval(prd)) +
  xlab(eval(prdcc)) +
  ggtitle("Index of Association p-values over sex rate and sample size", 
          subtitle = "with log-transformed axes") +
  labs(list(pch = "Mutation Rate")) +
  guides(pch = guide_legend(override.aes = list(alpha = 1))) +
  theme(legend.position = "bottom") +
  asp(1)
```

```{r, rd_sexrate_pval, fig.width = 10, fig.height = 10}
ggplot(vals, aes(x = sexrate, y = rbarD, color = log(p.rD))) +
  geom_jitter(alpha = 0.25, height = 0) +
  geom_boxplot(color = "black", notch = TRUE, width = 0.5, alpha = 0.25) +
  log_rd_color + 
  facet_grid(sample~mutation_rate) +
  labs(list(
    x = "Rate of Sexual Reproduction",
    y = eval(rd),
    color = eval(prd),
    title = "Detection of clonal reproduction",
    subtitle = "for even and uneven mutation rates")
    )
```


```{r, clone-correction-significance, fig.width = 10}
ggplot(even_mutation, aes(x = rbarD, y = rbarDcc, color = significance)) +
  geom_point(alpha = 0.25) +
  geom_abline(slope = 1) +
  facet_grid(sample~sexrate) +
  scale_x_continuous(breaks = c(0, 0.5, 1)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(override.aes = list(alpha = 1))) +
  asp(1) +
  labs(list(
    x = eval(rd),
    y = eval(rdcc),
    title = "Differences in probability due to clone correction", 
    subtitle = "20 loci; even mutation rate")
  ) 
ggplot(odd_mutation, aes(x = rbarD, y = rbarDcc, color = significance)) +
  geom_point(alpha = 0.25) +
  geom_abline(slope = 1) +
  facet_grid(sample~sexrate) +
  scale_x_continuous(breaks = c(0, 0.5, 1)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(override.aes = list(alpha = 1))) +
  asp(1) +
  labs(list(
    x = eval(rd),
    y = eval(rdcc),
    title = "Differences in probability due to clone correction", 
    subtitle = "21 loci; uneven mutation rate")
  ) 
```


```{r, Hexp_rbarD, fig.width = 10}
ggplot(vals, aes(x = Hexp, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  log_rd_color +
  facet_wrap(~mutation_rate) +
  labs(list(
    title = "Index of association vs. Gene Diversity",
    subtitle = "for even and uneven mutation rates",
    y = eval(rd),
    x = eval(Hexp),
    color = eval(prd))
    )
```

```{r, Hexp_rbarD_facet, fig.width = 10}
ggplot(even_mutation, aes(x = Hexp, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  log_rd_color + 
  asp(1) +
  facet_grid(sample~sexrate) +
  labs(list(
    title = "Index of association vs. Gene Diversity",
    subtitle = "by sex rate and sample size for 20 loci; even mutation rate",
    y = eval(rd),
    x = eval(Hexp),
    color = eval(prd))
    )
ggplot(odd_mutation, aes(x = Hexp, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  log_rd_color + 
  asp(1) +
  facet_grid(sample~sexrate) +
  labs(list(
    title = "Index of association vs. Gene Diversity",
    subtitle = "by sex rate and sample size for 21 loci; uneven mutation rate",
    y = eval(rd),
    x = eval(Hexp),
    color = eval(prd))
    )
```

```{r, gE5_lE5, fig.width = 10}
ggplot(even_mutation, aes(x = Evenness, y = E.5, color = rbarD)) +
  geom_point(alpha = 0.25) +
  scale_color_viridis() +
  geom_abline(slope = 1) +
  asp(0.75) +
  facet_grid(sample~sexrate) +
  xlab(eval(E5a)) +
  ylab(eval(E5g))
ggplot(odd_mutation, aes(x = Evenness, y = E.5, color = rbarD)) +
  geom_point(alpha = 0.25) +
  scale_color_viridis() +
  geom_abline(slope = 1) +
  asp(0.75) +
  facet_grid(sample~sexrate) +
  xlab(eval(E5a)) +
  ylab(eval(E5g))
```

```{r, gE5_rbarD, fig.width = 10}
ggplot(even_mutation, aes(x = E.5, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  log_rd_color +
  asp(0.75) +
  facet_grid(sample~sexrate) +
  xlab(eval(E5g)) +
  ylab(eval(rd))
ggplot(odd_mutation, aes(x = E.5, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  log_rd_color +
  asp(0.75) +
  facet_grid(sample~sexrate) +
  xlab(eval(E5g)) +
  ylab(eval(rd))
```


```{r, eveness_rd, fig.height = 10, fig.width = 10}
ggplot(filter(even_mutation, p.rD >= 0.05), aes(x = Evennesscc, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  # scale_color_viridis(option = "magma", direction = -1,
  #                     breaks = log(c(2, 5, 10, 25, 50, 100)),
  #                     labels = c(2, 5, 10, 25, 50, 100)) +
  log_rd_color + 
  facet_grid(sexrate~sample) +
  theme_dark() +
  geom_vline(xintercept = max((even_mutation %>% filter(sexrate == "1.0000"))$Evennesscc), lty = 2) +
  asp(0.5) + 
  ggtitle(expression(paste(bar(r)[d], " vs. mean allele evenness")), 
          subtitle = "For p >= 0.05 with even mutation rates") +
  ylab(expression(bar(r)[d])) +
  xlab(expression(paste(E[5], " (averaged over 20 loci)")))
ggplot(filter(odd_mutation, p.rD >= 0.05), aes(x = Evennesscc, y = rbarD, color = log(p.rD))) +
  geom_point(alpha = 0.25) +
  # scale_color_viridis(option = "magma", direction = -1,
  #                     breaks = log(c(2, 5, 10, 25, 50, 100)),
  #                     labels = c(2, 5, 10, 25, 50, 100)) +
  log_rd_color + 
  facet_grid(sexrate~sample) +
  theme_dark() +
  geom_vline(xintercept = max((odd_mutation %>% filter(sexrate == "1.0000"))$Evennesscc), lty = 2) +
  asp(0.5) + 
  ggtitle(expression(paste(bar(r)[d], " vs. mean allele evenness")), 
          subtitle = "For p >= 0.05 with uneven mutation rates") +
  ylab(expression(bar(r)[d])) +
  xlab(expression(paste(E[5], " (averaged over 21 loci)")))
```



```{r, grabbag}
ggplot(vals, aes(x = sexrate, y = E.5, fill = sample)) +
  geom_boxplot() +
  facet_wrap(~mutation_rate)
ggplot(vals, aes(x = sexrate, y = H, fill = sample)) +
  geom_boxplot() +
  facet_wrap(~mutation_rate)
ggplot(vals, aes(x = sexrate, y = B, fill = sample)) +
  geom_boxplot() +
  facet_wrap(~mutation_rate)
```


```{r, CF_B}
ggplot(vals, aes(x = CF, y = B, color = sample)) +
  geom_point() +
  facet_grid(sample~mutation_rate)
```

