---
title: "ROC Analysis"
author: "Zhian N. Kamvar"
date: "October 16, 2016"
output:
  html_document: default
  html_notebook: default
---

## Introduction

I've created 10,000 simulations of populations undergoing different levels of
sexual reproduction and analyzed the index of association along with genotypic
and allelic diversity metrics, including an analysis of the contribution of each
locus to the genotypic diversity. 

This document will perform a power analysis of $\bar{r}_d$ by calculating
Receiver Operator Characteristic (ROC) Curves. These curves are useful in
assessing the power of an analysis by assessing the true positive fraction and
false positive fraction against different values of $\alpha$. 

In our case, we are asking the question:

> How well can $\bar{r}_d$ detect non-random mating (clonal reproduction)?


## Analysis

### Overall ROC Curve

We will be looking at factors such as sample size, percent of clonal 
reproduction, and variable mutation rates. To perform the analyis, the data will
be split up into two sets: "Random Mating" and "Non-Random Mating". The Random 
mating set will always be the data set where the sex rate is one. This will
serve to assess our Type 1 (False Positive) fraction. The Non-Random mating
component will each rate of sexual reproduction less than one. In this data set,
that means that 9 total ROC curves will be produced for each overall comparison.

### ROC Curve by Seed

Since we simulated the populations in a way such that each parent population is
one of 10 replicates spawned from 100 unique seeds across rates of sexual
reproduction, we can group the populations by seed and calculate an ROC Curve
for each seed.

### Area Under the ROC Curve (AURC)

The Area under the ROC Curve will be calcuated via the `auc()` function via
*flux*. 

## Setup

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width = 10, fig.height = 10)
```


```{r, required_packages}
library('zksimanalysis')
```

We load the data previously processed in `ROC_calculation.Rmd`.

```{r, load_data, cache = TRUE, results = "hide"}
system.time({
res <- load("../../simulation_results/ROC_data.rda")
})
res
```


```{r, generate_sexrate}
alpha <- seq(0, 1, by = 0.01)
old_theme <- theme_set(theme_bw())
old_theme <- theme_update(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

## SSR Data

### Overall ROC 

```{r, curves, fig.width = 10, fig.height = 10}
oroc_plot <- ggplot(roc_overall, aes(x = `False Positive`, y = `True Positive`)) +
  geom_ribbon(aes(ymin = `True Positive` - TPsd, ymax = `True Positive` + TPsd, fill = sample), 
              alpha = 0.25) +
  geom_line(aes(color = sample)) +
  facet_grid(sexrate ~ mutation_rate + clone_correction) +
  geom_abline(slope = 1, lty = 3) +
  labs(list(
    # title = "ROC Curves",
    # subtitle = "over mutation rate and clone-correction by sex rate",
    pch = "Sample Size",
    color = "Sample Size",
    fill = "Sample Size"
  )) +
  theme(aspect.ratio = 1) +
  theme(legend.position = "top") +
  theme(panel.grid.minor = element_blank()) +
  theme(text = element_text(size = 14)) +
  scale_color_brewer(palette = "Set1")
oroc_plot
total_oroc_plot <- ggplot(total_roc, aes(x = `False Positive`, y = `True Positive`)) +
  geom_ribbon(aes(ymin = `True Positive` - TPsd, ymax = `True Positive` + TPsd, fill = sample), 
              alpha = 0.25) +
  geom_line(aes(color = sample)) +
  facet_grid(mutation_rate ~ clone_correction) +
  geom_abline(slope = 1, lty = 3) +
  labs(list(
    title = "ROC Curves",
    pch = "Sample Size",
    color = "Sample Size",
    fill = "Sample Size"
  )) +
  theme(aspect.ratio = 1) +
  theme(legend.position = "top") +
  theme(text = element_text(size = 14)) +
  scale_color_brewer(palette = "Set1")
total_oroc_plot
```

### AURC

Now we can visualize the data as facetted boxplots

```{r, AURC_plot, fig.width = 10, fig.height = 10}

# Add an annotation indicating that 0.5 is random assignment.
text_annotation <- data.frame(x = 0, y = 0.5, text = "0.5 = random", 
                              sample = "n = 10", mutation_rate = "even")

ggplot(AURC_by_seed, aes(x = factor(sexrate), y = AURC)) +
  geom_boxplot(aes(fill = clone_correction)) +
  annotate("rect", xmin = 0, xmax = 9.75, ymin = 0, ymax = 0.5, alpha = .2) +
  geom_boxplot(aes(fill = clone_correction)) +
  facet_grid(sample~mutation_rate) +
  geom_text(aes(x = x, y = y, label = text), hjust = -0.1, vjust = 1.5, color = "gray25",
            data = text_annotation) +
  geom_point(aes(x = factor(sexrate), y = AURC, pch = clone_correction), data = AURC_overall,
             position = position_dodge(width = 0.75), color = "black", fill = "white") +
  scale_shape_manual(values = c(21, 24)) + 
  labs(list(
    fill = "Clone Correction",
    pch = "Overall",
    x = "Rate of Sexual Reproduction",
    title = expression(paste("Area Under the ROC Curve for ", bar(r)[d])),
    subtitle = "calculated over sample size and mutation rate"
  )) +
  scale_fill_brewer(palette = "Set2") +
  theme(text = element_text(size = 14))


```



### ANOVA

There is a lot of data here, and it's clear that there is a difference between
clone-corrected data and uncorrected data, but to be sure of that, we'll do
ANOVA tests between each pair of clone-corrected and un-corrected data, grouped
by sex rate, sample size, and mutation rate.

```{r, fig.width=10, fig.height = 7}
AURC_by_seed %>% 
  group_by(sexrate, mutation_rate, sample) %>% 
  do(aov(AURC ~ clone_correction, data = .) %>% broom::tidy()) %>% 
  filter(term != "Residuals") %>% 
  select(p.value) %>% 
  ungroup() %>%
  mutate(mutation_rate = factor(mutation_rate, labels = c("Even Mutation Rate", "Uneven Mutation Rate"))) %>% 
  ggplot(aes(x = sexrate, y = sample, fill = p.value)) + 
  geom_tile() + 
  geom_text(aes(label = round(p.value, 2)), color = "grey50") + 
  scale_fill_viridis(direction = -1, begin = 0, end = 1, guide = "legend",
                     breaks = c(0, 0.01, 0.05, 0.1, 1)) + 
  facet_wrap(~mutation_rate, nrow = 1) + 
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme(aspect.ratio = 4/9) +
  theme(legend.position = "bottom") +
  theme(strip.background = element_blank()) +
  theme(strip.text = element_text(size = 14, face = "bold")) +
  # theme(panel.spacing = unit(0, "cm")) +
  theme(axis.title.x = element_blank()) +
  theme(axis.text.x = element_blank()) +
  theme(axis.ticks.x = element_blank()) +
  theme(text = element_text(size = 14)) +
  theme(plot.margin = unit(c(1, 1, 0, 1), "lines")) +
  labs(list(
    # title = "ANOVA over Clone Correction",
    x = "Rate of Sexual Reproduction",
    y = "Sample Size",
    fill = "p-value (Clone Correction)"
  )) -> cc_anova
AURC_by_seed %>% 
  group_by(sexrate, mutation_rate, clone_correction) %>% 
  do(aov(AURC ~ sample, data = .) %>% broom::tidy()) %>% 
  filter(term != "Residuals") %>% 
  select(p.value) %>% 
  ggplot(aes(x = sexrate, y = clone_correction, fill = p.value)) + 
  geom_tile() + 
  geom_text(aes(label = round(p.value, 2)), color = "grey50") + 
  scale_fill_viridis(direction = -1, begin = 0, end = 1, guide = "legend",
                     breaks = c(0, 0.01, 0.05, 0.1, 1), option = "B") + 
  facet_wrap(~mutation_rate, nrow = 1) + 
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme(text = element_text(size = 14)) +
  theme(aspect.ratio = 2/9) +
  theme(legend.position = "top") +
  theme(strip.background = element_blank()) +
  theme(strip.text = element_blank()) +
  # theme(panel.spacing = unit(0, "cm")) +
  labs(list(
    # title = "ANOVA over Sample Size",
    x = "Rate of Sexual Reproduction",
    y = "Clone Correction",
    fill = "p-value (Sample Size)       "
  )) -> size_anova

library(gtable)
library(grid)
g1 <- ggplotGrob(cc_anova)
# g1 <- gtable_add_cols(cc_anova, unit(0,"mm")) # add a column for missing legend
g2 <- ggplotGrob(size_anova)
g <- rbind(g1, g2, size="first") # stack the two plots
g$widths <- unit.pmax(g1$widths, g2$widths) # use the largest widths
# center the legend vertically
# g$layout[grepl("guide", g$layout$name),c("t","b")] <- c(1,nrow(g))
grid.newpage()
grid.draw(g)
```


## ROC Table

```{r total_roc_table, fig.height=5, fig.width=8}
total_roc_table <- total_roc %>% filter(alpha %in% c(0.01, 0.05))
total_roc_table %>% 
  select(clone_correction, mutation_rate, sample, alpha, `True Positive`) %>% 
  spread(sample, `True Positive`) %>%
  arrange(alpha, mutation_rate, clone_correction)
  ggplot(total_roc_table, aes(x = sample, y = `True Positive`, color = `False Positive`, pch = factor(alpha))) +
    geom_point(size = 3, position = position_dodge(width = 0.5)) + 
    # geom_linerange(aes(ymax = `True Positive` + TPsd, ymin = `True Positive` - TPsd),
    #                position = position_dodge(width = 0.5), color = "black") +
    scale_color_viridis() +
    scale_y_continuous(limits = c(0, 1)) +
    facet_grid(mutation_rate ~ clone_correction) +
    theme(aspect.ratio = 0.5) +
    theme(legend.position = "bottom") +
    labs(list(
      title = "Overall power to detect non-random mating",
      shape = expression(alpha),
      color = "False Positive Rate",
      x = "Sample Size",
      y = "True Positive Rate"
    ))
    
```

```{r, fig.height=6, fig.width=9}
overall_roc_table <- roc_overall %>% 
  filter(alpha %in% c(0.01)) %>%
  mutate(clone_correction = factor(clone_correction, labels = c("clone-corrected", "whole data"))) %>%
  mutate(mutation_rate = factor(mutation_rate, labels = c("Even Mutation Rate", "Uneven Mutation Rate")))
powerplot <- overall_roc_table %>% 
  ggplot(aes(y = `True Positive`, x = sexrate, color = sample, alpha = `False Positive`)) +
  geom_point(size = 2) + 
  # geom_errorbar(aes(ymax = `True Positive` + TPsd, ymin = `True Positive` - TPsd), width = 0.25) +
  geom_line(aes(group = sample)) +
  facet_grid(mutation_rate ~ clone_correction) +
  scale_alpha(range = c(1, 0.5)) +
  theme(aspect.ratio = 0.5) +
  theme(strip.text = element_text(size = 12, face = "bold")) +
  theme(strip.background = element_blank()) +
  theme(text = element_text(size = 14)) +
  theme(legend.position = "bottom") +
  theme(legend.box = "vertical") +
  labs(list(
    title = "Power to detect non-random mating (p = 0.01)",
    x = "Rate of Sexual Reproduction",
    y = "True Positive Rate",
    alpha = "False Positive Rate",
    color = "Sample Size"
  ))
powerplot
```


## Genomic Data


```{r genomic_curve}
g_oroc_plot <- ggplot(g_roc_overall, aes(x = `False Positive`, y = `True Positive`)) +
  geom_ribbon(aes(ymin = `True Positive` - TPsd, ymax = `True Positive` + TPsd, fill = sample), 
              alpha = 0.25) +
  geom_line(aes(color = sample)) +
  facet_wrap(~sexrate) +
  geom_abline(slope = 1, lty = 3) +
  labs(list(
    # title = "ROC Curves",
    # subtitle = "over mutation rate and clone-correction by sex rate",
    pch = "Sample Size",
    color = "Sample Size",
    fill = "Sample Size"
  )) +
  theme(aspect.ratio = 1) +
  theme(legend.position = "top") +
  theme(panel.grid.minor = element_blank()) +
  theme(text = element_text(size = 14)) +
  scale_color_brewer(palette = "Set1")
g_oroc_plot
```

```{r genomic_curve_total, fig.height=5, fig.width=5}
g_total_oroc_plot <- ggplot(g_total_roc, aes(x = `False Positive`, y = `True Positive`)) +
  geom_ribbon(aes(ymin = `True Positive` - TPsd, ymax = `True Positive` + TPsd, fill = sample), 
              alpha = 0.25) +
  geom_line(aes(color = sample)) +
  geom_abline(slope = 1, lty = 3) +
  labs(list(
    pch = "Sample Size",
    color = "Sample Size",
    fill = "Sample Size"
  )) +
  theme(aspect.ratio = 1) +
  theme(legend.position = "top") +
  theme(text = element_text(size = 14)) +
  scale_color_brewer(palette = "Set1")
g_total_oroc_plot
```



```{r, fig.height=7, fig.width=7}
# Add an annotation indicating that 0.5 is random assignment.
text_annotation <- data.frame(x = 0, y = 0.5, text = "0.5 = random", 
                              sample = "n = 10", mutation_rate = "even")

g_AURC_by_seed_plot <- ggplot(g_AURC_by_seed, aes(x = factor(sexrate), y = AURC)) +
  geom_point(position = position_jitter(height = 0, width = 0.125), alpha = 0.5) +
  annotate("rect", xmin = 0, xmax = 9.75, ymin = 0, ymax = 0.5, alpha = 0.2) +
  facet_wrap(~sample) +
  geom_text(aes(x = x, y = y, label = text), hjust = -0.1, vjust = 1.5, color = "gray25",
            data = text_annotation) +
  geom_point(aes(x = factor(sexrate), y = AURC, fill = "white"), data = g_AURC_overall, 
             pch = 21, position = position_dodge(width = 0.75), color = "black", 
             size = 3, alpha = 0.75) +
  labs(list(
    x = "Rate of Sexual Reproduction",
    title = expression(paste("Area Under the ROC Curve for ", bar(r)[d])),
    subtitle = "calculated over sample size"
  )) +
  scale_fill_manual(values = "white", labels = "Overall") +
  theme(legend.title = element_blank()) +
  theme(legend.position = "bottom") +
  theme(text = element_text(size = 14)) +
  theme(strip.text = element_text(size = 14, face = "bold")) +
  theme(aspect.ratio = 0.75)
g_AURC_by_seed_plot
```


## Session Information

```{r session_info}
options(width = 100)
devtools::session_info()
```
